### PR TITLE
[6/N] Logs: add logger per context

### PR DESCRIPTION
**Pull request by lusmoura**
Thank you for contributing to the Cohere Toolkit!

- [ ] **PR title**: "area: description"
  - Where "area" is whichever of interface, frontend, model, tools, backend, etc. is being modified. Use "docs: ..." for purely docs changes, "infra: ..." for CI changes.
  - Example: "deployment: add Azure model option"

- [ ] **PR message**: ***Delete this entire checklist*** and replace with
    - **Description:** a description of the change
    - **Issue:** the issue # it fixes, if applicable
    - **Dependencies:** any dependencies required for this change

- [ ] **Add tests and docs**: Please include testing and documentation for your changes
- [ ] **Lint and test**: Run `make lint` and `make run-tests`

**AI Description**

The changes in this PR update the logger setup and usage across multiple files.

- The get_logger function is replaced with a LoggerFactory class, which encapsulates the logger creation logic.
- The logger strategy and level are now configurable, with the default strategy set to "structlog" and the level set to "INFO".
- The logger is now accessed through the `ctx` object, using the `get_logger` method.
- The logger setup and usage have been updated in various modules, including `chat`, `config`, `routers`, `services`, and `tests`.
- The BaseLogger class and its methods have been modified to accept keyword arguments (`**kwargs`) instead of individual parameters.
- The get_context_log function now accepts any context object and removes specific keys from the context dictionary before returning it.
- The StructuredLogging class has been updated to accept a renderer argument, allowing for console or JSON output.
- The LoggingMiddleware class now retrieves the logger using the `ctx` object.

### CODE REVIEW

File: src/backend/services/logger/strategies/structured_log.py [hunk 1]
```diff
@@ -14,29 +17,43 @@ def wrapper(self, **kwargs):
...
+
+
 class StructuredLogging(BaseLogger):
     def __init__(self, level: str = "info", renderer: str = "json"):
         self.setup(level, renderer)
         self.logger = structlog.get_logger()
 
     def setup(self, level: str = "info", renderer: str = "json"):
+        level = getattr(logging, level.upper(), None)
+        # logging.basicConfig(level=level)
```
Comment (tianjing-li) [inline src/backend/services/logger/strategies/structured_log.py:50]: Should we remove this comment?

File: src/backend/services/logger/strategies/structured_log.py [hunk 2]
```diff
@@ -58,7 +75,7 @@ def setup(self, level: str = "info", renderer: str = "json"):
         structlog.configure(
             processors=processors,
             wrapper_class=structlog.make_filtering_bound_logger(level),
-            cache_logger_on_first_use=True,  # Remove this line to make changes to the logger
+            # cache_logger_on_first_use=True,  # Remove this line to make changes to the logger
```
Comment (tianjing-li) [inline src/backend/services/logger/strategies/structured_log.py:78]: Same here

**Review by tianjing-li — APPROVED** [review]
LGTM - just 2 minor comments

**Review by scott-cohere — APPROVED** [review]
lgtm, just remove those comments as TJ mentioned.

**Pull request merged**